build.assets/robotest/Makefile

```diff
@@ -0,0 +1,326 @@
+	mkdir -p $(ROBOTEST_BUILDDIR)
+
+$(ROBOTEST_UPGRADE_IMAGE_CACHE):
+	mkdir -p $(ROBOTEST_UPGRADE_IMAGE_CACHE)
+
+$(ROBOTEST_IMAGEDIR):
+	mkdir -p $(ROBOTEST_IMAGEDIR)
+
+$(ROBOTEST_TELE_CACHE):
+	mkdir -p $(ROBOTEST_TELE_CACHE)
+
+$(ROBOTEST_BUILD_TMP):
+	mkdir -p $(ROBOTEST_BUILD_TMP)
+
+$(ROBOTEST_CACHE_TMP):
+	mkdir -p $(ROBOTEST_CACHE_TMP)
+
+.PHONY:docker-image
+docker-image: ## Create robotest docker image with the current user/group baked in.
+docker-image: $(ROBOTEST_DOCKER_IMAGE_IIDFILE)
+
+$(ROBOTEST_DOCKER_IMAGE_IIDFILE): Dockerfile Makefile
+	docker build \
+		--build-arg BASE=$(ROBOTEST_DOCKER_UPSTREAM_IMAGE) \
+		--build-arg UID=$(shell id -u) \
+		--build-arg GID=$(shell id -g) \
+		--iidfile $(ROBOTEST_DOCKER_IMAGE_IIDFILE) \
+		--tag $(ROBOTEST_DOCKER_IMAGE) .
+
+.PHONY: images
+images: ## Create all images necessary to run robotest CI tests.
+images: image-robotest images-upgrade image-opscenter
+
+.PHONY: image-robotest
+image-robotest: ## Create robotest cluster image using the current tele.
+image-robotest: $(ROBOTEST_IMAGE)
+
+$(ROBOTEST_IMAGE): $(TELE_OUT) $(ROBOTEST_IMAGE_SRC) | $(ROBOTEST_IMAGEDIR) $(ROBOTEST_BUILD_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_BUILD_TMP)))
+	$(eval TGT := $(TMP)/robotest-$(GRAVITY_VERSION).tar)
+	$(TELE_OUT) build \
+		$(ROBOTEST_APP_YAML) \
+		--state-dir=$(PACKAGES_DIR) \
+		--version $(GRAVITY_VERSION) \
+		--output=$(TGT)
+	mv $(TGT) $(ROBOTEST_IMAGE)
+	rm -rf $(TMP)
+
+.PHONY: images-upgrade
+images-upgrade: ## Create/fetch robotest cluster images needed for upgrade testing.
+images-upgrade:
+	$(MAKE) -j $(ROBOTEST_UPGRADE_IMAGES)
+
+.PHONY: image-opscenter
+image-opscenter: ## Move opscenter.tar cluster image into place for test execution.
+image-opscenter: $(ROBOTEST_OPSCENTER_IMAGE)
+
+$(ROBOTEST_OPSCENTER_IMAGE): $(OPSCENTER_OUT) | $(ROBOTEST_IMAGEDIR)
+	cp $(OPSCENTER_OUT) $(ROBOTEST_OPSCENTER_IMAGE)
+
+# The following rule populates the build specific docker image mount
+# from cached images. Fulfills the $(ROBOTEST_UPGRADE_IMAGES) targets.
+# .PRECIOUS because these need to be present through the run target.
+.PRECIOUS: $(ROBOTEST_IMAGEDIR)/robotest-%.tar
+$(ROBOTEST_IMAGEDIR)/robotest-%.tar: $(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar | $(ROBOTEST_IMAGEDIR) $(ROBOTEST_BUILD_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_BUILD_TMP)))
+	$(eval ATOMIC := $(TMP)/robotest-$*.tar)
+	cp $< $(ATOMIC)
+	mv $(ATOMIC) $@
+	rm -rf $(TMP)
+
+# The following target is a pattern rule for populating the upgrade image cache
+# Fulfills the $(ROBOTEST_UPGRADE_CACHE_IMAGES) targets.
+# .PRECIOUS because it is a cache shared across versions.
+.PRECIOUS: $(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar
+$(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar: $(ROBOTEST_TELE_CACHE)/tele-% $(ROBOTEST_UPGRADE_BASE_IMAGE_SRC) | $(ROBOTEST_UPGRADE_IMAGE_CACHE) $(ROBOTEST_CACHE_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_CACHE_TMP)))
+	$(eval TGT := $(TMP)/robotest-$*.tar)
+	$< build \
```

Comment:
Should be fixed by 5c47ccf. :crossed_fingers:

Reply:
Unfortunately 5c47ccf failed because not all upgrade from `tele` binaries have the fix from #2023, leading to `[ERROR]: no credentials for https://get.gravitational.io:443` if `--hub` is blanket applied to old tele builds.

I'd rather not branch the upgrade base tele build logic at this point, so I've opted to containerize all tele builds to isolate them from ~jenkins state.  We'll see how this goes shortly.

Reply:
Refactored to sandbox tele builds within a docker container in c91653e.  This prevents any possible contamination from ~jenkins.  It took a couple more patches to get the new dockerized builds working, but now that they are (f86dc73) we don't see this issue any more.